windows/lcow_runtime.go

```diff
@@ -0,0 +1,426 @@
+// +build windows
+
+package windows
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/Microsoft/hcsshim"
+	"github.com/boltdb/bolt"
+	eventsapi "github.com/containerd/containerd/api/services/events/v1"
+	containerdtypes "github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/events"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/plugin"
+	"github.com/containerd/containerd/runtime"
+	"github.com/containerd/containerd/typeurl"
+	"github.com/containerd/containerd/windows/hcsshimopts"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
+const (
+	lcowRuntimeName = "lcow"
+)
+
+var (
+	lcowPluginID = fmt.Sprintf("%s.%s", plugin.RuntimePlugin, lcowRuntimeName)
+)
+
+var _ = (runtime.Runtime)(&LCOWRuntime{})
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		ID:   lcowRuntimeName,
+		Type: plugin.RuntimePlugin,
+		Init: NewLCOWPlugin,
+		Requires: []plugin.PluginType{
+			plugin.MetadataPlugin,
+		},
+	})
+}
+
+func NewLCOWPlugin(ic *plugin.InitContext) (interface{}, error) {
+	if err := os.MkdirAll(ic.Root, 0700); err != nil {
+		return nil, errors.Wrapf(err, "could not create state directory at %s", ic.Root)
+	}
+
+	m, err := ic.Get(plugin.MetadataPlugin)
```

Comment:
If this plugin just wants to use boltdb, it does not need to use the metadata bolt database. The metadata bolt database is primarily for storing related and similar objects. It is also intended to be used with the `metadata` package for operating on the database. Each plugin can have its own database and allows easier cleanup of that plugin. For example, each snapshotter has its own boltdb database, with its own structure. The metadata database tends to follow the structure `v1.<namespace>.<object type>.<object-name>`

Reply:
I think he just duplicated what I initially did with the windows runtime, my bad, didn't realize at the time :).

Regarding the use of metadata, it should disappear once the snapshotter for windows is added, I don't think there's anything that'll need to be stored by the windows runtimes.

Reply:
Yes, this part is duplicated from the Windows runtime. I can make a followup to fix this in both Windows and LCOW runtimes if that works, or just remove it as snapshotters are added.